<a href="https://colab.research.google.com/github/KBE25/hotel_sentiment/blob/main/notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Overview

# Business Understanding

# Data Understanding

# Data Preparation

Data Preparation will contain the following parts:

## Data Acquisition

In [1]:
import kagglehub
import pandas as pd
import os


In [2]:
# Download latest dataset version
path = kagglehub.dataset_download("thedevastator/booking-com-hotel-reviews")

# Identify the path for the dataset
print("Path to dataset files:", path)

Path to dataset files: /kaggle/input/booking-com-hotel-reviews


In [3]:
# The path where the data is in the environment
base_dataset_directory = "/kaggle/input/booking-com-hotel-reviews"
csv_file_name = "booking_reviews copy.csv"
full_csv_path = os.path.join(base_dataset_directory, csv_file_name)
df = pd.read_csv(full_csv_path)


## Exploratory Data Analysis

In [8]:
df.shape

(26675, 16)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26675 entries, 0 to 26674
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   index            26675 non-null  int64  
 1   review_title     26674 non-null  object 
 2   reviewed_at      26570 non-null  object 
 3   reviewed_by      26570 non-null  object 
 4   images           938 non-null    object 
 5   crawled_at       26386 non-null  object 
 6   url              26386 non-null  object 
 7   hotel_name       26386 non-null  object 
 8   hotel_url        26386 non-null  object 
 9   avg_rating       26386 non-null  float64
 10  nationality      26370 non-null  object 
 11  rating           26386 non-null  float64
 12  review_text      26386 non-null  object 
 13  raw_review_text  26202 non-null  object 
 14  tags             26202 non-null  object 
 15  meta             26202 non-null  object 
dtypes: float64(2), int64(1), object(13)
memory usage: 3.3+ MB


In [6]:
df.describe()

index    avg_rating        rating
count  26675.000000  26386.000000  26386.000000
mean   13337.000000      8.449242      8.560669
std     7700.553552      0.716427      1.566554
min        0.000000      3.800000      1.000000
25%     6668.500000      8.100000      7.900000
50%    13337.000000      8.500000      9.000000
75%    20005.500000      8.900000     10.000000
max    26674.000000     10.000000     10.000000

Below I will be exploring some of the rows from the Datafram to determine which of the columns are relevant for the analysis:

In [9]:
df.head(-5)

index                                       review_title  \
0          0                                        Exceptional   
1          1  I highly recommend this b&b! We enjoyed it a lot!   
2          2                                        Exceptional   
3          3  My stay in the house was a experiencing bliss ...   
4          4  One bedroom apartment with wonderful view and ...   
...      ...                                                ...   
26665  26665                                             Superb   
26666  26666                                           Fabulous   
26667  26667                                        Exceptional   
26668  26668                                             Superb   
26669  26669        Almost perfect hotel - criticisms are minor   

             reviewed_at reviewed_by  \
0           11 July 2021      Kyrylo   
1       24 November 2019     Dimitri   
2         3 January 2020    Virginia   
3       8 September 2019      Kannan   
4           23 June 2019         Sue   
...                  ...         ...   
26665  14 September 2018         Ann   
26666     1 October 2018       Hamed   
26667      17 March 2020        Olya   
26668     8 October 2019      Subodh   
26669    18 October 2018   Anonymous   

                                                  images  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4      https://cf.bstatic.com/xdata/images/xphoto/squ...   
...                                                  ...   
26665                                                NaN   
26666                                                NaN   
26667                                                NaN   
26668                                                NaN   
26669                                                NaN   

                 crawled_at  \
0      07/20/2021, 01:09:27   
1      07/20/2021, 01:09:27   
2      07/20/2021, 01:09:27   
3      07/20/2021, 01:09:28   
4      07/20/2021, 01:09:28   
...                     ...   
26665  07/20/2021, 02:15:50   
26666  07/20/2021, 02:15:50   
26667  07/20/2021, 02:15:50   
26668  07/20/2021, 02:15:50   
26669  07/20/2021, 02:15:50   

                                                     url  \
0      https://www.booking.com/reviews/be/hotel/villa...   
1      https://www.booking.com/reviews/be/hotel/villa...   
2      https://www.booking.com/reviews/be/hotel/hydro...   
3      https://www.booking.com/reviews/be/hotel/villa...   
4      https://www.booking.com/reviews/be/hotel/hydro...   
...                                                  ...   
26665  https://www.booking.com/reviews/be/hotel/oudhu...   
26666  https://www.booking.com/reviews/be/hotel/warwi...   
26667  https://www.booking.com/reviews/be/hotel/winds...   
26668  https://www.booking.com/reviews/be/hotel/best-...   
26669  https://www.booking.com/reviews/be/hotel/amigo...   

                           hotel_name  \
0                     Villa Pura Vida   
1                     Villa Pura Vida   
2              Hydro Palace Apartment   
3                     Villa Pura Vida   
4              Hydro Palace Apartment   
...                               ...   
26665                 Martin's Relais   
26666  Warwick Brussels - Grand Place   
26667                  Logies Windsor   
26668                  Hotel Ter Elst   
26669         Rocco Forte Hotel Amigo   

                                               hotel_url  avg_rating  \
0      https://www.booking.com/hotel/be/villa-pura-vi...         9.7   
1      https://www.booking.com/hotel/be/villa-pura-vi...         9.7   
2      https://www.booking.com/hotel/be/hydro-palace....         9.2   
3      https://www.booking.com/hotel/be/villa-pura-vi...         9.7   
4      https://www.booking.com/hotel/be/hydro-palace.

In [11]:
# Handle potential NaN values for comparison between raw_review_text and review_text
# For this I fill NaN with a placeholder string to avoid issues, as NaN != NaN
df_temp = df.fillna({'review_text': '', 'raw_review_text': ''}).copy()

# Count how many rows have identical content
identical_reviews = (df_temp['review_text'] == df_temp['raw_review_text']).sum()
# Only count rows where both are non-null
total_reviews_comparable = len(df_temp) - df_temp['review_text'].isnull().sum() - df_temp['raw_review_text'].isnull().sum()

print(f"Number of reviews where 'review_text' and 'raw_review_text' are identical (after filling NaN): {identical_reviews}")
print(f"Total comparable reviews (where both are non-null): {total_reviews_comparable}")
print(f"Percentage identical: {((identical_reviews / total_reviews_comparable) * 100):.2f}%")

Number of reviews where 'review_text' and 'raw_review_text' are identical (after filling NaN): 289
Total comparable reviews (where both are non-null): 26675
Percentage identical: 1.08%


Given that only 1.08% of review_text and raw_review_text are identical,this indicates significant differences between the two. Because of this it's crucial to understand the nature of these disparities before deciding which column to retain for further analysis.

In [14]:
# Look at descriptive statistics of lengths
print("\nDescriptive statistics for 'review_text' length:")
print(df_temp['review_text_len'].describe())
print("\nDescriptive statistics for 'raw_review_text' length:")
print(df_temp['raw_review_text_len'].describe())

# Check for rows where one is significantly longer than the other (e.g., >20 characters difference)
print("\n--- Examples where review_text is MUCH longer (diff > 20 chars) ---")
longer_review_text_diff = df_temp[df_temp['review_text_len'] - df_temp['raw_review_text_len'] > 20]
print(f"Number of such rows: {len(longer_review_text_diff)}")
if not longer_review_text_diff.empty:
    print(longer_review_text_diff[['review_text', 'raw_review_text']].head(3).to_string())

print("\n--- Examples where raw_review_text is MUCH longer (diff > 20 chars) ---")
longer_raw_review_text_diff = df_temp[df_temp['raw_review_text_len'] - df_temp['review_text_len'] > 20]
print(f"Number of such rows: {len(longer_raw_review_text_diff)}")
if not longer_raw_review_text_diff.empty:
    print(longer_raw_review_text_diff[['review_text', 'raw_review_text']].head(3).to_string())


Descriptive statistics for 'review_text' length:
count    26675.000000
mean       160.166111
std        201.930996
min          0.000000
25%         47.000000
50%         78.000000
75%        204.000000
max       3368.000000
Name: review_text_len, dtype: float64

Descriptive statistics for 'raw_review_text' length:
count    26675.000000
mean       782.758163
std        547.427211
min          0.000000
25%         83.000000
50%        837.000000
75%       1193.000000
max       4392.000000
Name: raw_review_text_len, dtype: float64

--- Examples where review_text is MUCH longer (diff > 20 chars) ---
Number of such rows: 149
                                                                                           review_text raw_review_text
54   Very clean and warm chalet. Friendly hosts willing to help and provide everything that is needed.                
75                                                                           Vicinity to train station                
241          

In [15]:
# Find a few rows where they are different (and both non-empty) for manual inspection
diff_reviews_sample = df_temp[(df_temp['review_text'] != df_temp['raw_review_text']) &
                              (df_temp['review_text'] != '') & (df_temp['raw_review_text'] != '')].sample(min(5, len(df_temp))).copy()

print("\n--- Detailed Examples of Differences (5 random samples) ---")
for index, row in diff_reviews_sample.iterrows():
    print(f"\nRow Index: {index}")
    print(f"review_text:     '{row['review_text']}'")
    print(f"raw_review_text: '{row['raw_review_text']}'")
    print("-" * 50)


--- Detailed Examples of Differences (5 random samples) ---

Row Index: 8560
review_text:     'There are no comments available for this review'
raw_review_text: '<p class="review_none"><em>There are no comments available for this review</em></p>'
--------------------------------------------------

Row Index: 9534
review_text:     'Bring a warm pyjama in the Ardennes!,

Single room for one small but lovely duvet, everything you need in the bathroom, ample space in cupboards and wardrobes, super cosy restaurant, nice staff.'
raw_review_text: '<p class="review_neg "><svg aria-label="Negative point" class="bk-icon -fonticon-removecircle review_item_icon" fill="#BDBDBD" height="14" role="img" width="14" viewbox="0 0 32 32" aria-hidden="true" focusable="false"><path d="m16 0c-8.84 0-16 7.16-16 16s7.16 16 16 16 16-7.16 16-16-7.16-16-16-16zm8 17.6h-16v-3.2h16z"></path></svg><span itemprop="reviewBody">Bring a warm pyjama in the Ardennes!</span></p>,<p class="review_pos "><svg aria-label="Posi

Based on the above information, I can conclude the following:

*   raw_review_text is significantly longer for the vast majority of reviews.
*   raw_review_text contains HTML tags and possibly meta-information like itemprop="reviewBody", and even indicators of positive/negative aspects (review_pos, review_neg).
*   review_text appears to be a cleaned-up, extracted version of the core text content from within raw_review_text, maybe even combining multiple aspects into one string.

The conclusion then is to keep raw_review_text and drop review_text because raw_review_text offers completeness, providing more original content. This gives more control over preprocessing, allowing me to precisely handle HTML tags and embedded sentiment indicators. This granular control is crucial for extracting richer features, as raw_review_text contains valuable review_pos/review_neg classes essential for advanced sentiment or aspect-based analysis.

Further evaluating the 'meta' column to see if it contains relevant information for the analysis:

In [10]:
# Displaying the first few non-null entries
print("First 10 non-null entries in 'meta' column:")
print(df['meta'].dropna().head(10).tolist())

# Displaying a random sample
print("\n10 random non-null entries in 'meta' column:")
print(df['meta'].dropna().sample(10).tolist())

First 10 non-null entries in 'meta' column:
["{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}"]

10 random non-null entries in 'meta' column:
["{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https://www.booking.com'}", "{'language': 'en-gb', 'source': 'https:

Based on the above checks, I will be dropping the following columns that are not relevant for this analysis:

1. index: Redundant DataFrame index.

2. images: Too sparse, and image analysis is out of scope for this analysis as we are using text for hotel rating prediction.

3. crawled_at: Irrelevant timestamp for this analysis as this is when the review was crawled from the site.

4. url: Unique identifier for the review page, no predictive power.

5. hotel_url: Unique identifier for the hotel page, but this becomes redundant if hotel_name or hotel_id is used.

6. meta: content is consistently redundant, specifying 'en-gb' language and 'booking.com' source, and therefore provides no unique or discriminative information for this analysis.

7. review_text: as it seems to be a simplified version of raw_review_text and is better to keep the columns with full information to better extract richer features for the analysis.

In [17]:
# Dropping unnecesary columns
# Create a list of columns to drop
columns_to_drop = ['index', 'images', 'crawled_at', 'url', 'hotel_url', 'meta', 'review_text']

# Perform the drop operation
df_cleaned = df.drop(columns=columns_to_drop, errors='ignore')

# Inspecting new dataframe
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26675 entries, 0 to 26674
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   review_title     26674 non-null  object 
 1   reviewed_at      26570 non-null  object 
 2   reviewed_by      26570 non-null  object 
 3   hotel_name       26386 non-null  object 
 4   avg_rating       26386 non-null  float64
 5   nationality      26370 non-null  object 
 6   rating           26386 non-null  float64
 7   raw_review_text  26202 non-null  object 
 8   tags             26202 non-null  object 
dtypes: float64(2), object(7)
memory usage: 1.8+ MB


# Exploratory Data Understanding